### PR TITLE
Add EnvVarHeaderContentGuard

### LINKED ISSUE #8007: Add EnvVarHeaderContentGuard for env-backed header secret validation
## Problem

`HeaderContentGuard` stores the expected secret in `header_value` on the guard record. When the shared secret is managed outside Pulp (injected by a reverse proxy, Kubernetes secret → pod env var, etc.), rotating it requires updating every guard that references that secret.

`HeaderContentGuard` expects the request header to be **Base64-encoded** and compares against the decoded value. That encoding keeps arbitrary UTF-8 secrets transport-safe over HTTP headers. We need the same wire-format behavior for env-backed secrets, while still reading the expected plaintext from `os.environ` at request time so rotation is a deployment/config change rather than a bulk API update across guard records.

## Proposed solution

Add an `EnvVarHeaderContentGuard` content guard type to pulpcore (alongside `HeaderContentGuard`):

| Field | Purpose |
|-------|---------|
| `header_name` | Request header to inspect (operator-configured) |
| `env_var` | Name of the environment variable holding the expected secret (plaintext UTF-8) |

**`permit()` behavior:**

1. Read `request.headers[header_name]`; deny if missing.
2. Base64-decode the header value; deny if not valid Base64.
3. Decode the result as UTF-8; deny if not valid UTF-8.
4. Read `os.environ[env_var]`; deny if unset or empty (after strip).
5. Compare decoded header bytes to the env value using `hmac.compare_digest` on UTF-8-encoded strings (supports non-ASCII secrets).
6. Deny on mismatch; allow otherwise.

Proxies send `base64(utf-8(secret))` in the header. The pod environment variable holds the **plaintext** secret.

**API:** New viewset at `/pulp/api/v3/contentguards/core/envvar_header/` (or equivalent `TYPE = "envvar_header"` naming), with RBAC/access policy matching `HeaderContentGuard`.

**v1 scope (suggested):**

- Base64-encoded header on the wire (plaintext secret in env var), matching `HeaderContentGuard` transport behavior.
- No `jq_filter` (can be added later if needed).
- No allowlist fields.

## Use case

Any deployment where:

- A trusted edge component injects a shared secret header on authorized requests.
- The secret is provisioned via environment variables (K8s `env`, Clowder, etc.).
- Operators want rotation without touching guard records in the database.
- Secrets may include non-ASCII UTF-8 characters when Base64-encoded on the wire.

## Alternatives considered

1. **Extend `HeaderContentGuard`** with an optional `env_var` field instead of a new type — fewer models, but mixes DB-stored and env-sourced secrets in one guard; harder to document and permission separately.
2. **Keep as a downstream plugin only** — we prototyped this in `pulp-service` (https://github.com/pulp/pulp-service/pull/1420); workable but duplicates a generic content-guard primitive that belongs in core.
3. **Continue using `HeaderContentGuard` + DB updates** — rotation requires updating every guard instance; does not scale for shared secrets across many distributions.
4. **Raw header compare (no Base64)** — simpler for ASCII-only secrets, but breaks non-ASCII secrets and is less consistent with `HeaderContentGuard` and HTTP proxy behavior.

## Acceptance criteria

- [ ] `EnvVarHeaderContentGuard` model, migration, serializer, viewset registered in pulpcore
- [ ] `permit()` Base64-decodes the header, then validates against `os.environ[env_var]` at request time
- [ ] Missing header, invalid Base64, invalid UTF-8, unset/empty env var, and wrong value → deny (403 on content app)
- [ ] Correct Base64-encoded header matching env plaintext → allow
- [ ] Unit tests for `permit()` edge cases (missing header, invalid Base64, wrong value, env unset/empty, trailing newline in env, non-ASCII UTF-8 secret)
- [ ] Functional test: guarded distribution denies without header, allows with matching Base64-encoded header when env is set
- [ ] Access policy registered for the new viewset (same pattern as `HeaderContentGuard`)
- [ ] OpenAPI / client bindings updated

## Related

- `HeaderContentGuard` — `pulpcore/app/models/publication.py`
- Hosted prototype: https://github.com/pulp/pulp-service/pull/1420 (PULP-2257) — implements Base64 decode + UTF-8 byte compare; can be adapted upstream or replaced once core ships this type
- Documentation: https://gitlab.cee.redhat.com/hosted-pulp/pulp-docs/-/blob/main/content-guards.md (Env Var Header Content Guard section)
- Related guard lifecycle discussion: #7990 (deleting guards in use)

### PR DESCRIPTION
**Pull request by decko**
## Summary
- Adds `EnvVarHeaderContentGuard` so a Base64-encoded request header can be checked against a content-app environment variable (`fixes #8007`). Rotation is a deploy/env change, not a DB update of `HeaderContentGuard.header_value`.
- `env_var` must be listed in `ENVVAR_HEADER_CONTENT_GUARD_ALLOWED_VARS` (default `[]`). The secret is never stored or returned by the API.
- Unit and functional tests use `ENVVAR_HEADER_GUARD_TEST_SECRET` from the Pulp container env (injected via `template_config.yml` / CI `pulp_env`).

## Test plan
- [ ] `pytest -k envvar_header` unit tests (secret + allowlist set in the pulp container)
- [ ] `pytest -k envvar_header` functional tests (create guard, 403 without/wrong header, 404 with matching Base64 header, 400 for disallowed `env_var`, access policy present)
- [ ] Confirm OpenAPI path `/pulp/api/v3/contentguards/core/envvar_header/` and that GET does not expose a secret value

Related: PULP-2257, pulp-service prototype https://github.com/pulp/pulp-service/pull/1420

Closes #8007 

Assisted-by: Cursor Grok 4.6

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: .github/workflows/scripts/before_install.sh [hunk 1]
```diff
@@ -49,8 +49,8 @@ plugin_name: "pulpcore"
 legacy_component_name: "pulpcore"
 component_name: "core"
 component_version: "${COMPONENT_VERSION}"
-pulp_env: {"PULP_CA_BUNDLE": "/etc/pulp/certs/pulp_webserver.crt"}
-pulp_settings: {"allowed_export_paths": ["/tmp"], "allowed_import_paths": ["/tmp"], "api_root": "/pulp/", "content_path_prefix": "/somewhere/else/", "csrf_trusted_origins": ["https://pulp:443"], "distributed_publication_retention_period": 3, "orphan_protection_time": 0, "task_diagnostics": ["memory"], "task_protection_time": 10, "tmpfile_protection_time": 10, "upload_protection_time": 10}
+pulp_env: {"ENVVAR_HEADER_GUARD_TEST_SECRET": "functional-test-secret-value", "PULP_CA_BUNDLE": "/etc/pulp/certs/pulp_webserver.crt"}
```
Comment (ggainey) [inline .github/workflows/scripts/before_install.sh]: the next plugin-template run will throw this away. Also - static secrets, even test ones, make gitleaks yell at us. Let's generate a random one "on the fly", something like this:

https://github.com/pulp/pulp-cli/pull/1447/changes#diff-91c0a0e51396a0be3f344ba35b45cc37d0f992032ff5b04938f2a8c296115931R72

Once we like it here, let's get this into plugin_template instead.

File: docs/user/guides/protect-content.md [hunk 2]
```diff
@@ -61,71 +61,46 @@ pulp content-guard header create --name header-guard --header-name X-Pulp-User -
...
-An explicitly provided `content_guard` on a distribution always takes precedence over the domain default.
+The environment variable name must be listed in
+`ENVVAR_HEADER_CONTENT_GUARD_ALLOWED_VARS`. Creating this guard is privileged: the name is a
+pointer into the content-app process environment, so this type should not be granted to
+untrusted tenants.
 
-!!! note
-    Setting or changing the domain default does not affect distributions that already exist — only new distributions created after the default is set will inherit it.
+Set the secret on every content-app replica (and typically the API as well). Setting it only on
+the API causes all content requests to be denied.
```
Comment (ggainey) [inline docs/user/guides/protect-content.md]:
```suggestion
the API causes all content requests to repositories using this kind of ContentGuard to be denied.
```

File: pulpcore/app/models/publication.py [hunk 3]
```diff
@@ -570,6 +557,77 @@ class Meta:
...
+            _logger.debug("Access not allowed - Header content is not Base64 encoded.")
+            raise PermissionError(_("Access denied.")) from None
+
+        try:
+            header_value = header_decoded_content.decode("utf-8")
+        except UnicodeDecodeError:
+            _logger.debug("Access not allowed - Header content is not valid UTF-8.")
+            raise PermissionError(_("Access denied.")) from None
+
+        expected = os.environ.get(self.env_var)
```
Comment (ggainey) [inline pulpcore/app/models/publication.py]: Should short-circuit prior to checking header-values - move to just after "not in allowed vars"

File: pulpcore/app/models/publication.py [hunk 4]
```diff
@@ -570,6 +557,77 @@ class Meta:
...
+    The expected secret is read from the content-app process environment at request time
+    so rotation only requires updating the environment and redeploying.
+    """
+
+    TYPE = "envvar_header"
+
+    header_name = models.TextField()
+    env_var = models.TextField()
+
+    def permit(self, request):
```
Comment (ggainey) [inline pulpcore/app/models/publication.py:578]: The error-msgs here fall into categories:
- pulp-configuration
  - allowed env-var, empty in env
- Forbidden 
  - disallowed env-var in header
  - env-var set, header contains it, header-value readable, header-val != env-var
- user header-configuration error
  - no header sent
  - unreadable header (b64, utf-8)

but all the user gets back is "Permission Error". 

I'd like to see a plain PermissionError("Access Denied") only for "Forbidden" and "pulp-configuration" (the user doesn't need to know the installation forgot to set a value - they just know that their 'secret' isn't valid for talking to this installation). 

For header-cfg-errors, we need to give the user info about what they did wrong - so a PermissionError() with the currently-debug-log string would be better - user can read, go "Oh, damn", fix their code, and try again.

File: pulpcore/app/models/publication.py [hunk 5]
```diff
@@ -570,6 +557,77 @@ class Meta:
...
+
+        try:
+            header_value = header_decoded_content.decode("utf-8")
+        except UnicodeDecodeError:
+            _logger.debug("Access not allowed - Header content is not valid UTF-8.")
+            raise PermissionError(_("Access denied.")) from None
+
+        expected = os.environ.get(self.env_var)
+        if expected is None or expected.rstrip("\r\n") == "":
+            _logger.warning(
```
Comment (ggainey) [inline pulpcore/app/models/publication.py]: Concur - warning for what is apparently a configuration-error. User doesn't need this, administrator does.  👍

File: pulpcore/app/settings.py [hunk 6]
```diff
@@ -321,6 +321,10 @@
 
 ALLOWED_EXPORT_PATHS = []
 
+# Process environment variable names EnvVarHeaderContentGuard may read.
+# Empty list means no variable is allowed.
+ENVVAR_HEADER_CONTENT_GUARD_ALLOWED_VARS = []
```
Comment (ggainey) [inline pulpcore/app/settings.py:326]: 👍

File: template_config.yml [hunk 7]
```diff
@@ -43,6 +43,7 @@ plugins:
     name: "pulp_certguard"
 pulp_env:
   PULP_CA_BUNDLE: "/etc/pulp/certs/pulp_webserver.crt"
+  ENVVAR_HEADER_GUARD_TEST_SECRET: "functional-test-secret-value"
```
Comment (ggainey) [inline template_config.yml]: As above - if we make this a randomized string, gitleaks is happier :)

**Review by ggainey — CHANGES_REQUESTED** [review]

File: docs/user/guides/protect-content.md (hunk 2 above)
Comment (decko) [reply docs/user/guides/protect-content.md]: Done.

File: .github/workflows/scripts/before_install.sh (hunk 1 above)
Comment (decko) [reply .github/workflows/scripts/before_install.sh]: Like this https://github.com/pulp/pulpcore/blob/14d2a0892/.github/workflows/scripts/before_install.sh#L44-L45 ?

File: pulpcore/app/models/publication.py (hunk 3 above)
Comment (decko) [reply pulpcore/app/models/publication.py]: Done.

File: pulpcore/app/models/publication.py (hunk 4 above)
Comment (decko) [reply pulpcore/app/models/publication.py:578]: Cool. I've worked it out and changed some log levels.

Comment (ggainey) [reply pulpcore/app/models/publication.py:578]: I thought about whether "env-var not in list" and "wrong secret" needed to be log.warning() or not. The advantage of the code as it is, is one can now detect - and *alert on* - someone attempting to brute-force a secret. We're good here.

File: .github/workflows/scripts/before_install.sh [hunk 8]
```diff
@@ -41,6 +41,8 @@ fi
 
 # Compose the scenario definition.
 mkdir -p .ci/ansible/vars
+ENVVAR_HEADER_GUARD_TEST_SECRET="$(python3 -c 'import secrets; print(secrets.token_urlsafe(50))')"
```
Comment (ggainey) [inline .github/workflows/scripts/before_install.sh:44]: 👍

**Review by ggainey — APPROVED** [review]
